### PR TITLE
187 implement rddl claim

### PR DESCRIPTION
**Pull request by LaurentMontBlanc**
# Description:
The RedeemClaim mechanic consists of the following 3 msg types:
- `CreateRedeemClaim` sent by participant to create a `RedeemClaim` entry
- `UpdateRedeemClaim` sent by validator to update the `RedeemClaim` entry with the liquid tx hash
- `ConfirmRedeemClaim` sent by service to update the `RedeemClaim` entry with the confirmation that RDDL have been sent to the supplied address

`RedeemClaim` entries can be queried by `beneficiary` + `id` (redeem claim id) and `liquidtxhash`.

The ante handler was updated so that `UpdateRedeemClaim` and `ConfirmRedeemClaim` can only be sent by the described actors.

`crddl` of the participant which sends a `CreateRedeemClaim` are burnt immediately so the same claim can not be redeemed twice.

## Changes:
- add RedeemClaim store
- add RedeemClaim msg server (create, update, confirm)
- add RedeemClaim ante handler
- adjust CheckValidatorDecorator for ante handler
- add e2e test case to suite
- add SendUpdateRedeemClaim to util
- add RedeemClaim queries (get by beneficiary and id / get by liquid tx hash)
- add burnClaimAmount to dao msgServer
- add claimAddress to dao params

### CODE REVIEW

File: x/dao/genesis.go [hunk 1]
```diff
@@ -8,6 +8,10 @@ import (
 
 // InitGenesis initializes the module's state from a provided genesis state.
 func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState) {
+	// Set all the redeemClaim
+	for _, elem := range genState.RedeemClaimList {
+		k.SetRedeemClaim(ctx, elem)
+	}
```
Comment (eckelj) [inline x/dao/genesis.go:14]: should be removed (see comment above).

File: x/dao/genesis.go [hunk 2]
```diff
@@ -17,6 +21,7 @@ func ExportGenesis(ctx sdk.Context, k keeper.Keeper) *types.GenesisState {
 	genesis := types.DefaultGenesis()
 	genesis.Params = k.GetParams(ctx)
 
+	genesis.RedeemClaimList = k.GetAllRedeemClaim(ctx)
```
Comment (eckelj) [inline x/dao/genesis.go:24]: should be removed (see comment above)

File: x/dao/genesis_test.go [hunk 3]
```diff
@@ -15,6 +15,16 @@ func TestGenesis(t *testing.T) {
 	genesisState := types.GenesisState{
 		Params: types.DefaultParams(),
 
+		RedeemClaimList: []types.RedeemClaim{
```
Comment (eckelj) [inline x/dao/genesis_test.go:18]: should be removed (see comment above)

File: x/dao/genesis_test.go [hunk 4]
```diff
@@ -26,5 +36,6 @@ func TestGenesis(t *testing.T) {
 	nullify.Fill(&genesisState)
 	nullify.Fill(got)
 
+	require.ElementsMatch(t, genesisState.RedeemClaimList, got.RedeemClaimList)
```
Comment (eckelj) [inline x/dao/genesis_test.go:39]: should be removed (see comment above)

File: x/dao/types/genesis.go [hunk 5]
```diff
@@ -10,6 +10,7 @@ const DefaultIndex uint64 = 1
 // DefaultGenesis returns the default genesis state
 func DefaultGenesis() *GenesisState {
 	return &GenesisState{
+		RedeemClaimList: []RedeemClaim{},
```
Comment (eckelj) [inline x/dao/types/genesis.go:13]: to be removed. see above

File: x/dao/types/genesis.go [hunk 6]
```diff
@@ -18,6 +19,16 @@ func DefaultGenesis() *GenesisState {
...
+	// Check for duplicated index in redeemClaim
+	redeemClaimIndexMap := make(map[string]struct{})
+
+	for _, elem := range gs.RedeemClaimList {
+		index := string(RedeemClaimKey(elem.Beneficiary, elem.Id))
+		if _, ok := redeemClaimIndexMap[index]; ok {
+			return fmt.Errorf("duplicated index for redeemClaim")
+		}
+		redeemClaimIndexMap[index] = struct{}{}
+	}
```
Comment (eckelj) [inline x/dao/types/genesis.go:31]: to be removed, see above

**Review by eckelj — CHANGES_REQUESTED** [review]

File: proto/planetmintgo/dao/genesis.proto [hunk 7]
```diff
@@ -1,12 +1,16 @@
...
 import "planetmintgo/dao/params.proto";
+import "planetmintgo/dao/redeem_claim.proto";
 
 option go_package = "github.com/planetmint/planetmint-go/x/dao/types";
 
 // GenesisState defines the dao module's genesis state.
 message GenesisState {
-  Params params = 1 [(gogoproto.nullable) = false];
+           Params      params          = 1 [(gogoproto.nullable) = false];
+  repeated RedeemClaim redeemClaimList = 2 [(gogoproto.nullable) = false];
```
[the remark this answers is not in the document]
Comment (LaurentMontBlanc) [reply proto/planetmintgo/dao/genesis.proto]: done

File: x/dao/genesis.go (hunk 1 above)
Comment (LaurentMontBlanc) [reply x/dao/genesis.go]: done

File: x/dao/genesis_test.go (hunk 3 above)
Comment (LaurentMontBlanc) [reply x/dao/genesis_test.go]: done

File: x/dao/genesis.go (hunk 2 above)
Comment (LaurentMontBlanc) [reply x/dao/genesis.go]: done

File: x/dao/types/genesis.go (hunk 5 above)
Comment (LaurentMontBlanc) [reply x/dao/types/genesis.go]: done

File: x/dao/types/genesis.go (hunk 6 above)
Comment (LaurentMontBlanc) [reply x/dao/types/genesis.go]: done

**Review by eckelj — APPROVED** [review]

**Pull request merged**